clang/docs/PointerAuthentication.rst

```diff
@@ -0,0 +1,265 @@
+
+- The address at which the pointer is stored in memory.  A pointer signed with a discriminator which incorporates its storage address is said to have **address diversity**.  In general, using address diversity means that a pointer cannot be reliably replaced by an attacker or used to reliably replace a different pointer.  However, an attacker may still be able to attack a larger call sequence if they can alter the address through which the pointer is accessed.  Furthermore, some situations cannot use address diversity because of language or other restrictions.
+
+- A constant integer, called a **constant discriminator**. A pointer signed with a non-zero constant discriminator is said to have **constant diversity**.  If the discriminator is specific to a single declaration, it is said to have **declaration diversity**; if the discriminator is specific to a type of value, it is said to have **type diversity**.  For example, C++ v-tables on arm64e sign their component functions using a hash of their method names and signatures, which provides declaration diversity; similarly, C++ member function pointers sign their invocation functions using a hash of the member pointer type, which provides type diversity.
+
+The implementation may need to restrict constant discriminators to be significantly smaller than the full size of a discriminator.  For example, on arm64e, constant discriminators are only 16-bit values.  This is believed to not significantly weaken the mitigation, since collisions remain uncommon.
+
+The algorithm for blending a constant discriminator with a storage address is implementation-defined.
+
+.. _Signing schemas:
+
+Signing schemas
+~~~~~~~~~~~~~~~
+
+Correct use of pointer authentication requires the signing code and the authenticating code to agree about the **signing schema** for the pointer:
+
+- the abstract signing key with which the pointer should be signed and
+- an algorithm for computing the discriminator.
+
+As described in the section above on `Discriminators`_, in most situations, the discriminator is produced by taking a constant discriminator and optionally blending it with the storage address of the pointer.  In these situations, the signing schema breaks down even more simply:
+
+- the abstract signing key,
+- a constant discriminator, and
+- whether to use address diversity.
+
+It is important that the signing schema be independently derived at all signing and authentication sites.  Preferably, the schema should be hard-coded everywhere it is needed, but at the very least, it must not be derived by inspecting information stored along with the pointer.  See the section on `Attacks on pointer authentication`_ for more information.
+
+Language Features
```

Comment:
Nit Pick. At this point, I was wondering, shouldn't this section say "C and C++ language features" rather than "language features"?
And then I noticed that this documentation lives in "clang/docs/...", so in a C and C++ (and Objective-C(++)) specific area.
Everything above this line in this document is, I think, mostly independent of source language. Which would make it feasible to split this documentation in a generic "LLVM" part (the part above this line); and a language-specific/frontend-specific part (the part below this line).
I'm not sure if that would make sense though - splitting the documentation may make it harder to follow and keep consistent?

Reply:
Yes, the llvm docs mainly focuses on IR definitions and backend details.  It points to this clang page otherwise, so it should be easily discoverable for other implementations.

Reply:
Right, this document is not meant to be LLVM documentation at all.  The part above describes the basic concepts of pointer authentication and is intentionally mostly divorced from how those concepts are presented in C/C++ or LLVM.  The LLVM docs link to it because you need to understand those basic concepts before you can understand the LLVM intrinsics.  It's roughly analogous to how you're expected to understand the basic concepts of throwing, unwinding, and catching exceptions before you go off to read about the LLVM `landingpad` instruction.